File: src/camera_cpp/src/camera_subscriber.cpp
#include "rclcpp/rclcpp.hpp"
#include "sensor_msgs/msg/image.hpp"
#include "cv_bridge/cv_bridge.h"
#include "opencv2/highgui/highgui.hpp"
#include "opencv2/imgproc/imgproc.hpp"

/*
29/7/26 Not used in this project, but can be used to subscribe to the camera image topic and display the images using OpenCV.
You will need to manually change the topic name in the code
*/



class ImageSubscriber : public rclcpp::Node
{
public:
    ImageSubscriber()
        : Node("image_subscriber")
    {
        rclcpp::QoS profile(rclcpp::KeepLast(10));
        profile.best_effort();
        // Create the subscription to the /camera/image topic
        subscription_ = this->create_subscription<sensor_msgs::msg::Image>(
            "/camera_0/image_raw",profile, std::bind(&ImageSubscriber::image_callback, this, std::placeholders::_1));
    }

private:
    void image_callback(const sensor_msgs::msg::Image::SharedPtr msg)
    {
        try
        {
            // Convert ROS image message to OpenCV image format (cv::Mat)
            cv_bridge::CvImagePtr cv_ptr = cv_bridge::toCvCopy(msg, sensor_msgs::image_encodings::BGR8);

            // Display the image using OpenCV
            cv::imshow("Camera Image", cv_ptr->image);
            cv::waitKey(1);  // wait for 1 ms
        }
        catch (const cv_bridge::Exception &e)
        {
            RCLCPP_ERROR(this->get_logger(), "Could not convert from '%s' to 'bgr8'.", msg->encoding.c_str());
        }
    }

    rclcpp::Subscription<sensor_msgs::msg::Image>::SharedPtr subscription_;

};

int main(int argc, char **argv)
{
    rclcpp::init(argc, argv);
    rclcpp::spin(std::make_shared<ImageSubscriber>());
    rclcpp::shutdown();
    return 0;
}

File: src/camera_py/camera_py/camera_subscriber.py
import rclpy
import rclpy.qos
from rclpy.node import Node
from sensor_msgs.msg import Image
from cv_bridge import CvBridge
import cv2



class CameraSubscriber(Node):
    def __init__(self):
        super().__init__('camera_subscriber')
        
        policy_best_effort = rclpy.qos.QoSProfile(reliability=rclpy.qos.ReliabilityPolicy.BEST_EFFORT,
                            durability = rclpy.qos.DurabilityPolicy.VOLATILE,
                            history=rclpy.qos.HistoryPolicy.KEEP_LAST,
                            depth = 10
                            )
        # Create a CvBridge instance to convert ROS images to OpenCV
        self.bridge = CvBridge()


        # Subscriber for the image topic, replace 'camera/image' with your topic name
        self.subscription = self.create_subscription(
            Image,
            '/camera/image',  # Adjust to your camera topic
            self.listener_callback,
            policy_best_effort
        )
        self.subscription  # prevent unused variable warning

    def listener_callback(self, msg):
        try:
            # Convert ROS Image message to OpenCV2 format
            cv_image = self.bridge.imgmsg_to_cv2(msg, desired_encoding='bgr8')

            # Display the image using OpenCV
            cv2.imshow("Camera Feed", cv_image)

            cv2.waitKey(1)  # Required to update the window and handle events
        except Exception as e:
            self.get_logger().error(f"Could not convert image: {e}")

def main(args=None):
    rclpy.init(args=args)
    node = CameraSubscriber()

    try:
        rclpy.spin(node)
    except KeyboardInterrupt:
        pass
    finally:
        # Clean up
        node.destroy_node()
        cv2.destroyAllWindows()  # Close OpenCV windows


if __name__ == '__main__':
    main()
